daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala

```diff
@@ -0,0 +1,197 @@
+          }
+    }
+
+  // undefined if `typ1` or `typ2` contains `TVar`, `TForAll`, or `TSynApp`.
+  private def compareType(typ1: Ast.Type, typ2: Ast.Type): Int =
+    compareType(0, FrontStack((typ1, typ2)))
+
+  private def compareText(text1: String, text2: String): Int =
+    Utf8.Ordering.compare(text1, text2)
+
+  private def isHexa(s: String): Boolean =
+    s.forall(x => '0' < x && x < '9' || 'a' < x && x < 'f')
+
+  @inline
+  private val underlyingCidDiscriminatorLength = 65
+
+  private def compareCid(cid1: Ref.ContractIdString, cid2: Ref.ContractIdString): Int = {
+    val lim = cid1.length min cid2.length
+
+    @tailrec
+    def lp(i: Int): Int =
+      if (i < lim) {
+        val x = cid1(i)
+        val y = cid2(i)
+        if (x != y) x compareTo y else lp(i + 1)
+      } else {
+        if (lim == underlyingCidDiscriminatorLength && (//
+          cid1.length != underlyingCidDiscriminatorLength && isHexa(cid2) || //
+          cid2.length != underlyingCidDiscriminatorLength && isHexa(cid1)))
+          throw new IllegalArgumentException(
+            "Try to a fresh contract id with  conflicting discriminators")
+        else
+          cid1.length compareTo cid2.length
+      }
+
+    lp(0)
+  }
+  @tailrec
+  // Only value of the same type can be compared.
+  private[this] def compareValue(x: Int, stack0: FrontStack[(SValue, SValue)]): Int =
+    stack0 match {
+      case FrontStack() =>
+        x
+      case FrontStackCons(tuple, stack) =>
+        if (x != 0) x
+        else
+          tuple match {
+            case (SUnit, SUnit) =>
+              compareValue(0, stack)
+            case (SBool(b1), SBool(b2)) =>
+              compareValue(b1 compareTo b2, stack)
+            case (SInt64(i1), SInt64(i2)) =>
+              compareValue(i1 compareTo i2, stack)
+            case (STimestamp(ts1), STimestamp(ts2)) =>
+              compareValue(ts1.micros compareTo ts2.micros, stack)
+            case (SDate(d1), SDate(d2)) =>
+              compareValue(d1.days compareTo d2.days, stack)
+            case (SNumeric(n1), SNumeric(n2)) =>
+              compareValue(n1 compareTo n2, stack)
+            case (SText(t1), SText(t2)) =>
+              compareValue(compareText(t1, t2), stack)
+            case (SParty(p1), SParty(p2)) =>
+              compareValue(compareText(p1, p2), stack)
+            case (SContractId(AbsoluteContractId(coid1)), SContractId(AbsoluteContractId(coid2))) =>
+              compareValue(compareCid(coid1, coid2), stack)
+            case (STypeRep(t1), STypeRep(t2)) =>
+              compareValue(compareType(t1, t2), stack)
+            case (SEnum(_, con1), SEnum(_, con2)) =>
+              compareValue((con1 compareTo con2), stack)
+            case (SRecord(_, _, args1), SRecord(_, _, args2)) =>
+              compareValue(0, zipAndPush(args1.asScala, args2.asScala, stack))
+            case (SVariant(_, con1, arg1), SVariant(_, con2, arg2)) =>
```

Comment:
Am I missing something here or are you comparing based on the constructor name? This doesn’t match the specification and it’s not an option since it would be inconsistent with the derived `Ord` instances you have in `DAML`.

Reply:
The first implementation will order using constructor name. 
This will allow to implement the rest of the stack (e.g. compiler) and start testing in parallel.   

Ordering with respect the order the constructor appears in the definition, is way more complicate and will be implement in a upcoming PR.

Reply:
Can we please add a big fat TODO for this or even better, error out on variants for now instead of doing the wrong thing?

Reply:
Good point. I added 2 TODOs